Extended and adapted to install PostgreSQL 10 for CentOS and Ubuntu

WolfgangPecho opened the pull request:
Tested with ansible 2.4.1 for Ubuntu 16.04 and CentOS-7.4

- replaced deprecated 'include:' by 'include_tasks:' for ansible 2.4.1
- added config and config template for postgresql 10
- adapted defaults.yml
- adapted templates/etc_systemd_system_postgresql.service.d_custom.conf.j2
  for PostgreSQL 10 the template outputs the wrong line:
	ExecStartPre=/usr/pgsql-10/bin/postgresql10-check-db-dir
  the correct one must include an additional '-':
	ExecStartPre=/usr/pgsql-10/bin/postgresql-10-check-db-dir

Hope you find it useful.

Anthony25 commented:
I tested on ArchLinux, and everything seems to work with postgresql 10.1.

benrudolph commented on defaults/main.yml:
```diff
@@ -319,6 +320,15 @@ postgresql_wal_receiver_timeout: 60s
+
+# These settings are ignored on a publisher.
+
+postgresql_max_logical_replication_workers: 4	# (>= 10) taken from max_worker_processes
```
The tabs here and on the changes throw a syntax error for me

WolfgangPecho replied:
Hi Ben,
I removed the tabs and hope it now works. Strangley on my system (Ubuntu 16.04/ansible 2.4.2.0) I hab no trouble.

WolfgangPecho commented:
Tabs removed.

benrudolph commented:
thanks @WolfgangPecho ! i think you accidentally closed this

sebalix commented on templates/HOWTO.postgresql.conf:
```diff
@@ -19,4 +21,6 @@ and save it under the 'templates' role directory
 5) If there are new options or some of them removed, update the 'default/main.yml' file and add a "(>= 9.X)" or "(<= 9.X)" comment to them.
 
-6) Update the '.travis.yml' file to test its new version.
+7) For yum based installation add version and minor version of postgresql in 'default/main.yml' under '# YUM settings' at end of file
```
You skipped the point 6) (5 to 7)

WolfgangPecho replied:
Hi Sébastien,
all ansible tests now pass including a new one for ansible 2.4.2.0/posgresql10.
Docker tests currently commented out - do not know docker ;-).

gclough replied:
I think that @sebalix was saying that in the file, there is now no step 6

1)
2)
3)
4)
5)
5)  (SHOULD THIS BE STEP 6?)
7)
8)

taybin commented:
When can this get merged in if it is passing the tests?

taybin commented:
Does this surplant #262 ?

WolfgangPecho commented:
@taybin: I compared the two pull requests (merifiiri and mine). Both do their job.

Differeneces are ( wp #267 <-> merifiri #262  ):

- defaults/main.yml:
  postgresql_version: 9.6                       <-> postgresql_version: 9.3
  postgresql_pgdg_releases uses 10:2 <-> postgresql_pgdg_releases uses 10:1

- tasks/main.yml and tasks/extension.yml:
  uses import_tasks:                 <-> uses include:

- templates/etc_systemd_system_postgresql.service.d_custom.conf.j2
  {% if postgresql_version == 10 %}  <-> {% if postgresql_version < 10 %}

- templates/postgresql.conf-10.orig
  the original postgresql.conf.sample <-> postgresql.conf with values set

Some idea what to do ?

taybin commented:
@WolfgangPecho Well, I like this one, personally, and would like to see it get merged!

gclough commented on .travis.yml:
```diff
@@ -43,8 +44,8 @@ script:
   - ./tests/idempotence_check.sh idempotence_out
 
   # Testing with docker (experimental)
```
If this is experimental @WolfgangPecho, then should we remove it and put it into a separate PR?

gclough commented on Vagrantfile:
```diff
@@ -7,8 +7,9 @@ Vagrant.configure('2') do |config|
   config.ssh.insert_key = false
   config.ssh.private_key_path = '~/.vagrant.d/insecure_private_key'
 
-  config.vm.define 'anxs' do |machine|
```
Is this related to the PostgreSQL v10 update?  It looks interesting, but isn't related to this Pull Request, right?

gclough commented on defaults/main.yml:
```diff
@@ -1,7 +1,7 @@
 
 # Basic settings
-postgresql_version: 9.3
+postgresql_version: 9.6            # if you use postgis too check whether the postgresql/postgis versions work together
```
Probably don't need that postgis comment in there.  There is a PR #310 , which will give reasonable defaults for PostGIS now.

gclough commented on defaults/main.yml:
```diff
@@ -117,6 +117,7 @@ postgresql_ssl_ciphers:
 postgresql_ssl_prefer_server_ciphers: on
 postgresql_ssl_ecdh_curve: 'prime256v1'
 postgresql_ssl_renegotiation_limit: 512MB # amount of data between renegotiations
+postgresal_ssl_dh_params_file: ''         # (>= 10)  specify file name for custom OpenSSL DH parameters
```
Typo in parameter  name `sal`... should be `sql`.

gclough commented on defaults/main.yml:
```diff
@@ -199,6 +200,7 @@ postgresql_bgwriter_flush_after:    0      # (>= 9.6) 0 disables,
 postgresql_effective_io_concurrency:        1   # 1-1000; 0 disables prefetching
 postgresql_max_worker_processes:            8   # (change requires restart)
 postgresql_max_parallel_workers_per_gather: 0   # (>= 9.6) taken from max_worker_processes
+postgresql_max_parallel_workers:            8   # (>= 9.6)
```
This is >=10... not 9.6

WolfgangPecho replied:
Fixed

gclough commented on defaults/main.yml:
```diff
@@ -347,7 +358,9 @@ postgresql_cpu_index_tuple_cost:       0.005   # same scale as above
 postgresql_parallel_tuple_cost:        0.1     # same scale as above (>= 9.6)
 postgresql_parallel_setup_cost:        1000.0  # same scale as above (>= 9.6)
-postgresql_min_parallel_relation_size: 8MB     # (>= 9.6)
+postgresql_min_parallel_relation_size: 8MB     # (= 9.6) replaced by below two parameters in 10
```
Do we need the comment about "replaced below"?  I'm not sure if that matches the style of the others, which normally just list > < = and the version, etc.

WolfgangPecho replied:
It may be helpful as a reminder if transitioning from 9.6 to 10.

gclough commented on defaults/main.yml:
```diff
@@ -587,7 +602,8 @@ postgresql_vacuum_multixact_freeze_table_age:     150000000   # (>= 9.3)
 postgresql_xmloption:    content
-postgresql_gin_fuzzy_search_limit:  0   # (<= 9.2)
+postgresql_gin_fuzzy_search_limit: 0     # (<= 9.2, >=10)
+postgresql_gin_pending_list_limit: 4MB   # (>= 10)
```
This was added in v9.5, not in v10.  I've added that to #314 too.

gclough commented on defaults/main.yml:
```diff
@@ -639,6 +655,11 @@ postgresql_max_locks_per_transaction: 64         # min 10
 postgresql_max_pred_locks_per_transaction: 64    # min 10
 
+postgresql_max_pred_locks_per_relation: -2       # (>= 10) negative values mean
+               # (max_pred_locks_per_transaction
```
Do we need these comments in this file?  If someone wants to know what it does, then they can check the doco, or the `postresql.conf` itself.

WolfgangPecho replied:
Right. Same for other comments taken from the original config file.  May be for less exerienced people like me helpful to read it here too :-).

gclough commented on tasks/extensions.yml:
```diff
@@ -2,7 +2,9 @@
 
 - include: extensions/contrib.yml
   when: postgresql_ext_install_contrib
+
```
Superseded by #293... but adding the spaces is probably a good idea.

gclough commented on templates/HOWTO.postgresql.conf:
```diff
@@ -2,11 +2,13 @@
 
-2) Extract the 'usr/share/postgresql/9.1/postgresql.conf.sample' file
-and save it under the 'templates' role directory
+2) Extract the 'usr/share/postgresql/9.1/postgresql.conf.sample' file with
```
Update to v10 here too, not v9.1

WolfgangPecho replied:
Done.

gclough commented on templates/HOWTO.postgresql.conf:
```diff
@@ -2,11 +2,13 @@
+   dpkg-deb -x postgresql-<version>_[...].deb dir/to/extract/to/
+
+   and save it under the 'templates' role directory
     => templates/postgresql.conf.9.{X}.orig
```
Update to `<version>` here too... but should we match the role and use `<postgresql_version>`?

WolfgangPecho replied:
Done.

gclough commented on templates/etc_systemd_system_postgresql.service.d_custom.conf.j2:
```diff
@@ -8,4 +8,8 @@ Group={{ postgresql_service_group }}
 Environment=PGDATA={{ postgresql_conf_directory }}
 
 ExecStartPre=
```
Why leave this blank line here... it's a historical typo, right?

WolfgangPecho replied:
Blank line deleted.

gclough commented on templates/etc_systemd_system_postgresql.service.d_custom.conf.j2:
```diff
@@ -8,4 +8,8 @@ Group={{ postgresql_service_group }}
 Environment=PGDATA={{ postgresql_conf_directory }}
 
 ExecStartPre=
+    {% if postgresql_version == 10 %}
```
At this point I'd say `>= 10`, as I suspect this format will stay for v11, v12, etc.

WolfgangPecho replied:
Done.

gclough replied:
Sorry for re-reviewing... but we should be using the `version_compare()` function here. (and anywhere else you're comparing versions)

` {% if postgresql_version | version_compare ('10', '>=') %}`

WolfgangPecho replied:
Done ;-)

gclough commented on templates/etc_systemd_system_postgresql.service.d_custom.conf.j2:
```diff
@@ -8,4 +8,8 @@ Group={{ postgresql_service_group }}
 ExecStartPre=
+    {% if postgresql_version == 10 %}
+ExecStartPre={{ postgresql_bin_directory }}/postgresql-{{ postgresql_version_terse }}-check-db-dir {{ postgresql_data_directory }}
+    {% else %}
```
Is the indentation on this correct?  Should it be better written in this format, ans using the `version_check()` builtin:
```
{% if postgresql_version | version_check('10', '>=') %}
  ExecStartPre={{ postgresql_bin_directory }}/postgresql-{{ postgresql_version_terse }}-check-db-dir {{ postgresql_data_directory }}
{% else %}
  ExecStartPre={{ postgresql_bin_directory }}/postgresql{{ postgresql_version_terse }}-check-db-dir {{ postgresql_data_directory }}
{% endif %}
```

WolfgangPecho replied:
Indentation removed. version_compare() filter used to check the version.

gclough commented on tests/playbook.yml:
```diff
@@ -6,4 +6,4 @@
     - ./vars.yml
   roles:
-    - postgresql
+    - postgresql 
```
No need for this extra space...

WolfgangPecho replied:
Removed.

gclough commented on tests/vars.yml:
```diff
@@ -1,14 +1,15 @@
 
 postgresql_users:
+
+  # postgresql 10 does not accepte unencrypted passwords 
```
I would change the comment to:

`+  # PostgreSQL >= 10 does not accept unencrypted passwords`

WolfgangPecho replied:
Done.

gclough commented on vagrant-inventory:
```diff
@@ -1,2 +1,3 @@
 [anxs]
-anxs.local ansible_ssh_host=192.168.88.22 ansible_ssh_port=22 ansible_ssh_user=vagrant ansible_ssh_private_key_file=~/.vagrant.d/insecure_private_key
+centos-vm ansible_ssh_host=192.168.88.22 ansible_ssh_port=22 ansible_ssh_user=vagrant ansible_ssh_private_key_file=~/.vagrant.d/insecure_private_key
```
Is this related to the v10 update?  It looks interesting, but should be in a different Pull Request.

WolfgangPecho replied:
I replaced this by vagrant-inventory/Vagrantfile of project's master branch.

gclough commented:
@taybin @WolfgangPecho , I've done a review, and left some comments in the code.  We really should get this merged, as v10 is coming into common use.  If you could review and update it, we would appreciate it.

gclough commented:
I've written #315 , which fixes some of the missing parameters.  Once this is merged @WolfgangPecho, then this Pull Request will need to be updated... because there are some clashes with the changes in there.

WolfgangPecho commented:
Only check for debian 9 failed. The others are ok.
Seems to be an issue with the setup:

TASK [Build or pull image if needed] *******************************************
fatal: [localhost]: FAILED! => {"changed": false, "msg": "Error building debian - code: None, message: error pulling image configuration: Get https://registry-1.docker.io/v2/library/debian/blobs/sha256:2b98c9851a37e18356abe407922726dc35c370f7cee76f435f94ee92950b4493: EOF, logs: [u'Step 1/6 : FROM debian:9\\n']"}

WolfgangPecho commented:
Added install/test for version 10 in docker environment.
Problem was  that posgresql does not accept unencrypted passwords. Changed 
tests/docker/group_vars/postgresql.yml to meet this requirement.

Note:  Forst try failed with a misleading error message

> TASK [ANXS.postgresql : PostgreSQL | Make sure the PostgreSQL users are present] ***
failed: [postgresql-10] (item=None) => {"censored": "the output has been hidden due to the fact that 'no_log: true' was specified for this result"}

Setting 'no_log: true' for a task seems to hide its error message if it fails. You have to remove no_log to see the error mesage.

Issue:
Currently there a two places where vars are defined for testing purposes:

- tests/docker/group_vars/postgresql.yml for docker based tests
- tests/vars.yml for vagrant based tests

We should have one location providing the variable settings for tests.

gclough commented:
@sebalix, and @UnderGreen ... how do we get this tested/verified, and merged?  It would be very nice to handle PostgreSQL v10.

gclough commented on tasks/users.yml:
```diff
@@ -13,7 +13,7 @@
     state: present
     login_user: "{{postgresql_admin_user}}"
-  no_log: true
+#  no_log: true
```
Rather than commenting it out, should we either set it to false, or remove the line entirely?

WolfgangPecho replied:
Reenabled. no_log: true suppresses the output of parameters used to  create the users. You may consider it a security feature.

gclough commented on templates/HOWTO.postgresql.conf:
```diff
@@ -2,21 +2,31 @@
 5) Update the new template following the major differences.
 
-5) If there are new options or some of them removed, update the 'default/main.yml' file and add a "(>= 9.X)" or "(<= 9.X)" comment to them.
+5) If there are new options or some of them removed, update the 'default/main.yml' file and add a "(>= 10)" or "(<= 10)" comment to them.
```
There are still two step 5's...

WolfgangPecho replied:
Fixed.

taybin commented:
I'm just excited to see this get so close to merging.  I'm not able to help test it at the moment though.

jlozadad commented:
im gonna see if i have some cycles to test on centos in the next week or so.

sebalix commented:
Just tested this on a Debian Stretch environment, it's working fine. It remains some glitch to fix, after that :+1:

WolfgangPecho commented:
Hmm ... ubuntu-upstart:14.04 failed. The others are ok. On my account all travis builds passed.

WolfgangPecho commented:
@gclough @jlozadad @taybin : Can one of you retrigger a travis build ?

jlozadad commented:
sorry, the only way I know is closing and reopening.

jlozadad commented:
looks like the build failed in
```
TASK [ANXS.postgresql : PostgreSQL | Add extensions to the databases] **********
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OperationalError: FATAL:  Peer authentication failed for user "postgres"
```

WolfgangPecho commented:
Yeap build failed. It's a bit strange. If I test with vagrant using  bento/ubunt-14.04 the tests/playbook.yml installs without any error. The travis job triggered on my github repo  passes
too.

WolfgangPecho commented:
Ok the result so far is:

- The error is reproducible for ANXS/postgresql travis build.
   The build for ubuntu-upsart:14.04 fails with

> TASK [ANXS.postgresql : PostgreSQL | Add extensions to the databases] **********
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OperationalError: FATAL:  Peer authentication failed for user "postgres"
failed: [postgresql-9.6] (item=({u'db': u'foobar'}, u'adminpack')) => {"changed": false, "item": [{"db": "foobar"}, "adminpack"], "msg": "unable to connect to database: FATAL:  Peer authentication failed for user \"postgres\"\n"}
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: OperationalError: FATAL:  Peer authentication failed for user "postgres"

- The travis build triggered by my fork passes.
- tests/playbook.yml  with vagrant using bento/ubuntu-14.04  installs without error

Have we a timing issue ? 
Is something else merged for the ANXS/postgresql build ?

jlozadad commented:
alright so I tested on centos 7 and it installed successfully. I ran again to check idempotent and 100% ok.

```
 ____________
< PLAY RECAP >
 ------------
        \   ^__^
         \  (oo)\_______
            (__)\       )\/\
                ||----w |
                ||     ||

anxs.local                 : ok=22   changed=0    unreachable=0    failed=0   

```

benrudolph commented:
🎉

Drewster727 commented:
🙌

WolfgangPecho commented:
Tnx to all who helped with this pull request.

Drewster727 commented:
Anyone get a chance to update the readme? Still shows 9.6 as the latest :) thanks for the excellent work!

jlozadad commented:
I merged a PR that updated it.

gclough commented:
@WolfgangPecho ... how did you fix this error:

> FATAL: Peer authentication failed for user "postgres"

I'm getting the same thing now on #291, and it doesn't make sense. :o(

gclough commented:
@Drewster727, and @jlozadad ... actually PR #308 was just to prepare the readme in expectation of this PR being merged.

Now that it's done, I've made #330 to show we've got support... and done some other tidy ups.

WolfgangPecho commented:
@gclough I could not fix the 'Perer authentification failed ...' error ;-). See my comment 
> Ok the result so far is: ...

5 days ago before jiozadad decided to merge the pull request. 
Maybe we should cancel the ubuntu14.04 test for travis ?

gclough commented:
@WolfgangPecho 
> Maybe we should cancel the ubuntu14.04 test for travis ?

No, it's working now.  I've committed #331 which fixes that.  If you have any easy steps to get Travis and Docker working, I'd certainly appreciate them!

greg.clough@ipreo.com
